### DISCUSSION TITLE
Tool length probing issue using "Manual touch off @ G59.3"

### DISCUSSION THREAD

**Discussion by cuishuang413** [terjeio/ioSender#255] [Q&A]
Hello,

As I'm learning ioSender, an issue has been troubling me for a bit. It is how to use Manual Touch Off @ G59.3.
I'm currently using the 2.0.41 XL verion of ioSender. My Gcode is all generated from Fusion 360. Tool change Gcodes are in the format such as T1M6 (change to tool 1).
Here is a discription of the issue I have:
1. I put the stock into the vise.
2. I home the machine. (the home button lit up in green)
3. I go to the G59.3 0,0,0 position and do a tool legth reference probe with the "Establish reference offset" box checked. After the probing completed, I got TLO ref'd lit up in red.
4. I find the stock origin in G54. I sometimes use a machenical edge finder or most time I just eyeball it. So I just move to where I want it to be 0,0,0 and I just hit the "Zero all" button.
5. I open the Gcode.
6. I hit cycle start.
7. Sometimes, the machine will move to probing position and ask for a probe. Why is this? Is it because I didn't choose the first tool of the Gcode when I did the reference probing?
8. Sometimes, the machine will not ask me to probe and do all the tool path of the first tool.
9. After the tool change code is up, for example, T2M6. The machine will go to G59.3 0,0,0 and ask for a probing.
10. I jog the machine until a few millimeters from the probe and send "$TPW".
11. The machine does a probe. The probing looks fine and complete. But there is no "TLO" showing next to the "TLO ref'd". In fact, there is no "TLO" box on my UI.
12. After the probe, I hit cycle start, the machine will move to the next path starting position but will not continue. In the "program" tab, there is a "*" in the T2M6 line. It looks like the M6 is still pending.

Above is the issue I have. I'm sure there is a very simple solution to this. But I have tried and I need help.

Thanks a lot!
Harry

**Comment by S2000Stefan** [comment]
Hello Harry,

> 11. The machine does a probe. The probing looks fine and complete. But there is no "TLO" showing next to the "TLO ref'd". In fact, there is no "TLO" box on my UI.

Please under Settings/General/10 for the parser status a tick must be set so that TLO is displayed. Can you try this out? 
Yours sincerely
 Stefan

**Reply by terjeio** [reply]
> Sometimes, the machine will move to probing position and ask for a probe. Why is this? Is it because I didn't choose the first tool of the Gcode when I did the reference probing?

Because the current tool is either unset (0) or set to a different tool than the first in the gcode file.
Before referencing the tool offset it is good practice to set the current tool with [M61](http://www.linuxcnc.org/docs/html/gcode/m-code.html#mcode:m61).

> After the probe, I hit cycle start, the machine will move to the next path starting position but will not continue. In the "program" tab, there is a "*" in the T2M6 line. It looks like the M6 is still pending.

Odd, perhaps duet to missing report elements, ref  answer from @S2000Stefan above. I am currently away on holiday so have no means to check this.

**Reply by cuishuang413** [reply]
Thanks for the quick response. I have checked the "parser status" and things changed a little. I now able to get the "TLO" show up in red after sending $TPW. However after the $TPW completed. The gcode still won't continue. The machine would move to the next position but will not continue.
I've attached a few screenshots here to help explaining my case.

![image](https://user-images.githubusercontent.com/16296237/204696770-0c73a05b-89ba-43de-9cfb-2429fe42c20d.png)

![image](https://user-images.githubusercontent.com/16296237/204696806-6bb1afca-ba37-400b-b11c-be626920cd41.png)

![image](https://user-images.githubusercontent.com/16296237/204696965-036da9a2-d7c9-4c01-b506-b2ad21328682.png)

**Reply by S2000Stefan** [reply]
Very strange, what comes to my mind now, is a current version of grblHAL on the controller?
I have made a video here of how I make my toolchange, maybe that will help you a bit.
You can play through it.
https://youtu.be/nU9RS5zZdo4
However, according to your list above, I can't see any errors.
 
@terjeio wish you a nice holiday.

greetings Stefan

**Reply by cuishuang413** [reply]
Here is some more inforamtion I hope to help identify the issue.
If I choose "Manual touch off" instead of "Manual touch off @ G59.3", the machine will stop and waiting for a "$TPW" at line S12000M3 instead of T2M6. I won't be able to do a touch off since the spindle is running. So it looks like the gcode is sent one line more than it should or grblHAL excuted one line than it should. I'm not knowledgeable enough to know the inner relation of them two.
Could it be a faulty touch probe? Noise in the probe signal line? I'm using this hardware and grblHAL firmware: https://github.com/phil-barrett/grblHAL-teensy-4.x
I hope this shows some other angles to this problem and maybe it would help to find the issue.

Thanks all.

**Reply by terjeio** [reply]
Sorry, this issue had fallen off my radar...

> the machine will stop and waiting for a "$TPW" at line S12000M3 instead of T2M6.

Add your failing gcode to a comment and I'll take a look.

**Reply by cuishuang413** [reply]
Thank you for your reply.
Here is an example:
```
(1001)
(T1  D=6 CR=0 - ZMIN=-13.2 - flat end mill)
(T2  D=8 CR=0 TAPER=90deg - ZMIN=-1 - spot drill)
G90 G94
G17
G21
G28 G91 Z0
G90

(2D Contour1)
T1 M6
S12000 M3
G54
M8
G0 X118.95 Y48.253
Z15
Z5
G1 Z2 F1006
Z1.3
G18 G2 X119.25 Z1 I0.3 K0 F3018.1
G1 X119.55
...
...
...
G0 Z15
M5
G28 G91 Z0
G90

(2D Chamfer1)
M9
T2 M6
S12000 M3
G54
M8
G0 X122.013 Y58.379
Z15
Z5
G1 Z2 F1338.1
Z-1
X122.737 Y58.04 F4014.2
G2 X125 Y59.478 I2.263 J-1.062
G1 X136
...
...
...
G0 Z15
M9
G28 G91 Z0
G90
G28 G91 X0 Y0
G90
M5
M30
```
The code is generated by Fusion 360 with the GRBL post processor. Since I wasn't able to figure out how to use the G59.3, I have been using my machine without tool change by asking the GRBL post processor to generator seperate files per tool. But I would still like to find a solution to this issue. I really appreciate your help.

**Reply by terjeio** [reply]
Seems to be working for me - there is a G18 that messes up the second tool change, I added a G17 after the G18 line to fix it.
This is where it stops for me:

![image](https://user-images.githubusercontent.com/20260062/215290149-f70fc5a4-806b-468b-a387-6ec27da94d43.png)

What is the controller version? Use _Help > About_ to find it.

**Reply by cuishuang413** [reply]
Wow! I think that's probably it! I won't be able to try it now. But I will update this once I do.

Thank you so much!

**Reply by terjeio** [reply]
> Wow! I think that's probably it!

I am not so sure, for me the "pending" status showed up in the correct line - for you it did not?
A possible reason is that the firmware version you are using is old and buggy.

grblHAL probes in the direction of the linear axis for tool changes per default, and the linear axis is changed by G17, G18 and G19. This was done to support lathe mode where the tool moves in the X plane. It is possible to lock the offset axis, and thus the direction, [here](https://github.com/grblHAL/core/blob/e2ba6db69cda9d7ab664c3d4bbce0c1737b81768/config.h#L200-L208) if you compile locally. This can also be done in the [Web Builder](http://svn.io-engineering.com:8080/) by selecting _Z-axis_ in the _TLO report_ drop down and flashing the generated binary:

![image](https://user-images.githubusercontent.com/20260062/215315783-415c4c1b-fd7a-4633-8912-7828c283e5d6.png)
